pkg/proxy/endpoints.go

```diff
@@ -109,12 +113,19 @@ func (info *BaseEndpointInfo) GetZoneHints() sets.Set[string] {
 
 // IP returns just the IP part of the endpoint, it's a part of proxy.Endpoint interface.
 func (info *BaseEndpointInfo) IP() string {
-	return proxyutil.IPPart(info.Endpoint)
+	if info.ip == "" {
+		info.ip = proxyutil.IPPart(info.Endpoint)
+	}
+	return info.ip
 }
 
 // Port returns just the Port part of the endpoint.
 func (info *BaseEndpointInfo) Port() (int, error) {
-	return proxyutil.PortPart(info.Endpoint)
+	var err error
+	if info.port <= 0 {
+		info.port, err = proxyutil.PortPart(info.Endpoint)
```

Comment:
likewise, we know the port at construct time; we can set `info.port` then rather than re-parsing it here (and thus we don't need the `error` return here)